### ISSUE TITLE
Add local gateway to allow browsers access to the network.

### ISSUE THREAD

**Issue by RangerMauve** [dat-land/dat-desktop#518]
The IPFS-companion extension has been doing amazing things for allowing browsers to integrate with the network.

They allow stuff like [rerouting ipfs:// URLs to the local gateway](https://github.com/ipfs-shipyard/ipfs-companion#automagical-detection-of-ipfs-resources) and providing a global [window.ipfs](https://github.com/ipfs-shipyard/ipfs-companion/issues/330) API that allows web pages to read and upload stuff to IPFS.

These features work by using an HTTP gateway for the local IPFS node installed on the user's machine which the extension will reroute requests to and proxy the global IPFS object API calls to.

Similar features in dat-land can only be found by installing the [Beaker](https://beakerbrowser.com) browser, which although it's really cool, it probably won't be the default choice for most casual users of the web.

Beaker supports the [dat://](https://beakerbrowser.com/docs/inside-beaker/dat-files-protocol.html) protocol and a [DatArchive](https://beakerbrowser.com/docs/apis/dat.html) global which could both be provided through a similar extension as IPFS-companion.

In order to replicate what IPFS-companion did for a dat-specific extension, the extension would need some way of accessing dat since we can't just run node in a browser extension.

If dat-desktop allowed users to access all the dat APIs from an HTTP gateway, all that people will need to access data from dat and use distributed web apps is to install dat-desktop and a browser extension for whatever browser they're already using.

Would PRs be welcome for this functionality?

**Comment by martinheidegger** [comment]
Having a browser plugin / integration that works out of the box is certainly a nice thing!

By skimming over the implementations of ipfs it didn't become clear to me what API's a companion extension would need. Do you have better insight? If not it might be very good to collect the data necessary.

Aside from that: It is my goal to disconnect dat-desktop from a dat-daemon that stores/manages the local dats. And it seems to me like it should be rather this deamon's job to provide the http API's.

**Comment by RangerMauve** [comment]
@martinheidegger The docs for the [IPFS HTTP API](https://ipfs.io/docs/api/) go over everything that's available, but it's essentially everything that the JS or Go APIs.

For the dat HTTP API, it should provide everything that the JS API provides. It should also have access control headers to only allow local connections to access it so that malicious web pages can't try to access it directly.

Would a PR against dat-daemon to add an HTTP API be welcome?

**Comment by martinheidegger** [comment]
`dat-daemon` is a few days old. It should provide all API's - at least the very basic ones (more complex stuff is still hidden - reading code right now to know more.

**Comment by karissa** [comment]
@RangerMauve yes!

**Comment by RangerMauve** [comment]
Looking at dat-daemon, I don't know if that's the right place to shove this.

I love the project, but it looks like it'd be better geared to managing archives rather than an RPC API that facilitates browser extensions.

I think that a good way forward would be to
- take [node-dat-archive](https://github.com/beakerbrowser/node-dat-archive) which already implements the DatArchive API from beaker
- wrap it in an RPC API (using websockets, but have it work over arbitrary duplex streams)
- embed the RPC into dat-desktop so that people can enable it with a flag.
- Having it just be standalone isn't good UX for casual users.
- Have an extension that will load dat URLs using the RPC API
- Have the extension provide a DatArchive API which uses an RPC to talk to the extension worker which then does access restrictions and relays RPC to dat-desktop

**Comment by RangerMauve** [comment]
@sammacbeth Any ideas so far for the DatArchive API?

How does your extension deal with setting the origin to be different between dats?

**Comment by sammacbeth** [comment]
I am currently working on a [prototype firefox extension](https://github.com/sammacbeth/dat-fox) to do just this, using dat-gateway to bridge http to dat. There is also [another project](https://github.com/HughIsaacs2/DatPart) doing the same for chrome.

In the extension's current state, it is possible to load dat sites, but I have not yet started on implementing the DatArchive API.

**Comment by sammacbeth** [comment]
@RangerMauve For the DatArchive API I'm currently testing [Native Messaging](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Native_messaging) for a private communication channel between the extension and node process. Injecting the API in the document can be done in the same way as the ipfs-companion extension.

Regarding setting origins correctly, one can use a [PAC file](https://github.com/sammacbeth/dat-fox/blob/master/pac.js) (via the extension proxy API) to implement this by proxying dat-like hostnames to the gateway, which then receives `GET http://{dat key}/path`, and can then return the correct document via dat.

**Comment by RangerMauve** [comment]
@sammacbeth I think I missed something, what does the origin get set to when I navigate to `dat://mydaturlhere`?

**Comment by RangerMauve** [comment]
@sammacbeth Sorry, I think I didn't explain myself well enough.
There were problems with gateways for stuff like IPFS in that all files served from the gateways were getting the same origins and therefore were able to access each other's cookies and localstorage.

With your approach, will different dat:// URLs be served under different origins, or will it just redirect to the gateway and in turn use the gateway's origin?

**Comment by sammacbeth** [comment]
@RangerMauve Yes, that is exactly the problem the extension solves vs. simply using the gateway server.

**Comment by RangerMauve** [comment]
That's awesome! Thank you for clarifying! :D

**Comment by RangerMauve** [comment]
Just gonna bump this since I've got the [dat-polyfill](https://github.com/RangerMauve/dat-polyfill) to a state where it can be used to run Fritter using a gateway.

Also, if there's a gateway set up in dat-desktop, you can use `registerProtocolHandler` to have dat-desktop intercept `dat://` urls, and have it change the url to point to the local gateway and open in the user's default browser.

That would enable people to just install the desktop and not worry about browser extension compatibility.

**Comment by sammacbeth** [comment]
I've put together a [test suite for DatArchive implementations](https://hashbase.io/sammacbeth/dat-archive-test), with the aim of helping the emerging implementations to converge and work with applications build for the Beaker browser reference implementation. It helped be find quite a few issues with the dat-fox implementation.

The source is available [here](https://github.com/sammacbeth/dat-archive-compat-tests) and I'm happy to take contributions to improve the coverage.

**Comment by RangerMauve** [comment]
@sammacbeth That's awesome! I'll make sure to try to run it with dat-polyfill when I've stabilized it!

What's funny is that Beaker itself is failing on the read as base64 and read as hex tests. :P

**Comment by soyuka** [comment]
This is good news! Adding the gateway to the desktop app would indeed be nice! 
If you want to work on this please base your work on the react branch :).

**Comment by RangerMauve** [comment]
Awesome, I've still got some stuff to do for dat-polyfill, but after I'm done with that I'll make this my next priority. :D

**Comment by martinheidegger** [comment]
I read into dat-polyfill and other solutions and my opinion is that we need a "service"/"manager"/"daemon" that allows to bridge the data towards a client.

@soyuka built https://github.com/soyuka/dat-daemon which is the closest to what I think would be a good idea. But it does have a few issues that make it hard to use with dat-desktop. I certainly would love to see a PR against https://github.com/dat-land/dat-desktop/pull/516 and https://github.com/dat-land/dat-desktop/pull/458 respectively.

**Comment by martinheidegger** [comment]
@RangerMauve My thinking is that dat-desktop should come with a daemon and start it if no other daemon is present but can fallback to another, locally running daemon. Maybe this could be the same with dat-cli?

The most straightforward way I can think is that the daemon communicates through a tcp port. As such it could be accessible through an extension _(even though it might not be available from a website)_. This could be used to add a dat from the browser directly show the continuous download progress of a dat (clicked in the browser).

**Comment by RangerMauve** [comment]
With regards to extensions, I think that this [libweb](https://github.com/mozilla/libdweb) work will replace the need for a gateway down the line, actually.
